src/main/java/org/grouplens/grapht/Instantiators.java

```diff
@@ -132,21 +144,25 @@ public Class getType() {
 
     private static class ProviderInstantiator implements Instantiator {
         private final Instantiator providerInstantiator;
-
         public ProviderInstantiator(Instantiator prov) {
             providerInstantiator = prov;
         }
 
         @Override
         public Object instantiate() throws ConstructionException {
+            Logger.trace(" provider invoked "+providerInstantiator);
             Provider<?> provider = (Provider) providerInstantiator.instantiate();
-            try {
+            try{
+                // --- push providerInstantiator to MDC ----------------------------------
+                mdcContext.put(provider.getClass().toString(), provider.toString());
```

Comment:
A few things here:

- The key for the MDC should be something static and meaningful; what this is supposed to say is 'the current provider is `provider`'. I would propose "org.grouplens.grapht.currentProvider". The value is fine as-is.
- We don't want to finish or close the context until after the provider is run (`provider.get()`). We are wanting to associate the provider's string representation with the key in the context for the duration of invoking the provider. So we want to:
    
    1. Create a context (and we need to create a new context here - we can't reuse the existing context)
    2. Run the provider
    3. Finish the context

    Finishing the context should be in a `finally` block, so it always happens.